File: 103.二叉树的锯齿形层序遍历.java
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import javax.swing.tree.TreeNode;

/*
 * @lc app=leetcode.cn id=103 lang=java
 *
 * [103] 二叉树的锯齿形层序遍历
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if(root==null){return null;}
        List<List<Integer>> result=new ArrayList<List<Integer>>();
        Queue<TreeNode> q=new LinkedList<TreeNode>();
        q.add(root);
        boolean isLeft=true;
        while(!q.isEmpty()){
            int size=q.size();
            Deque<Integer> r=new LinkedList<Integer>();
            for(int i=0;i<size;i++){
                TreeNode temp=q.poll();
                if(isLeft){
                    r.offerLast(temp.val);
                }
                else{
                    r.offerFirst(temp.val);
                }
                if(temp.left!=null){
                    q.add(temp.left);
                }
                if(temp.right!=null){
                    q.add(temp.right);
                }
            }
            result.add(new LinkedList<Integer>(r));
            isLeft=!isLeft;
        }
        return result;
    }
}
// @lc code=end

File: 103.二叉树的锯齿形层序遍历.py
'''
Author: your name
Date: 2021-04-20 21:39:09
LastEditTime: 2021-04-21 19:51:43
LastEditors: Please set LastEditors
Description: In User Settings Edit
FilePath: /leetcode/103.二叉树的锯齿形层序遍历.py
'''
#
# @lc app=leetcode.cn id=103 lang=python3
#
# [103] 二叉树的锯齿形层序遍历
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        nums=[]
        tree_queue=[]
        tree_queue.append(root)
        left=True
        while len(tree_queue)!=0:
            temp=[]
            size=len(tree_queue)
            for i in range(size):
                current_node=tree_queue.pop()
                if left:
                    temp.append(current_node.val)
                else:
                    temp.insert(0,current_node.val)
                if current_node.left:
                    tree_queue.insert(0,current_node.left)
                if current_node.right:
                    tree_queue.insert(0,current_node.right)
            left= not left
            nums.append(temp)
        return nums
                
# @lc code=end

